### PR TITLE
Student: view session results: do not show self responses #6290

### LINKED ISSUE #6290: Student: view session results: do not show self responses
Current:
When students view session results, they can see responses they gave to others. 

Suggestion:
Don't show responses the user gave to others. Instead, give a link to show the corresponding submission page. 

Justification:
Self-responses clutter the page. Some students get confused by self-responses.

### PR DESCRIPTION
**Pull request by dalessr**
Fixes #6290 #8497

**Outline of Solution**
- For each question, add "Your own responses:" notation before self-responses (the reason of using plural for `response` is because there could be multiple responses given to different receivers by this giver)
- Inside `StudentFeedbackResultsQuestionWithResponses`, separate `responseTables` to `selfResponseTables` and `othersResponseTables`

**Current:**
6290-current-2

6290-current-1

Status: ongoing (adding tests and modifying html files)

### CODE REVIEW

**Comment by dalessr** [comment]
Status: Ready for review

File: src/main/webapp/WEB-INF/tags/student/feedbackResults/questionWithResponses.tag [hunk 1]
```diff
@@ -12,8 +12,14 @@
 
     <c:if test="${questionWithResponses.questionDetails.individualResponsesShownToStudents}">
 
-      <c:forEach items="${questionWithResponses.responseTables}" var="responseTable">
-        <feedbackResults:responseTable responseTable="${responseTable}"/>
+      <c:forEach items="${questionWithResponses.othersResponseTables}" var="othersResponseTable">
+        <feedbackResults:responseTable responseTable="${othersResponseTable}"/>
+      </c:forEach>
+
+      <h4>Your own responses:</h4>
```
Comment (joanneong) [inline src/main/webapp/WEB-INF/tags/student/feedbackResults/questionWithResponses.tag]: I think it might be better to make this heading a little smaller than its current size. This is because right now it is the same size as the question, so it looks like this:

screen shot 2018-03-30 at 5 59 41 pm

But I think it can be slightly smaller to distinguish it further from the question.

**Comment by dalessr** [comment]
@joanneong Sure. How about this:
6290-current-4

6290-current-3

I used `<h5>` and `<b>` to make it clear since the font size is smaller.

Also, do you think it is OK to have nothing below `Your own responses` if there's no self-responses, like the example below?
6290-current-5

Or does it look better if there is no `Your own responses` notation when there's no self-responses?

**Comment by joanneong** [comment]
@dalessr Cool, I think `<h5>` and `<b>` work well :)

I thought about that as well when I was testing your branch. Personally I think it is okay because it's quite consistent like that, but maybe @xpdavid might have a different opinion? :P

**Comment by dalessr** [comment]
@joanneong sure 👌

**Comment by xpdavid** [comment]
@dalessr I have seen you already made the changes (Hide the heading if there is no response from you)
@joanneong Feel free to request review from me when ready.

**Comment by dalessr** [comment]
@joanneong Ready for review

File: src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java [hunk 2]
```diff
@@ -21,4 +22,28 @@ public FeedbackResultsQuestionDetails getQuestionDetails() {
         return responseTables;
```
Comment (xpdavid) [inline src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java]: Remove `getResponseTables` if it is not used.

File: src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java [hunk 3]
```diff
@@ -21,4 +22,28 @@ public FeedbackResultsQuestionDetails getQuestionDetails() {
...
     }
 
+    public List<FeedbackResultsResponseTable> getSelfResponseTables() {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java]: You can do some fancy stuff. :P. It's Java8 now.
```java
selfResponseTables = responseTables.stream().filter(responseTable -> responseTable.isGiverNameYou()).collect(Collectors.toList());
```

Also please add Javadoc for the method. A simple method summary is enough.

File: src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java [hunk 4]
```diff
@@ -21,4 +22,28 @@ public FeedbackResultsQuestionDetails getQuestionDetails() {
...
+        List<FeedbackResultsResponseTable> selfResponseTables = new ArrayList<>();
+        for (int i = 0; i < responseTables.size(); i++) {
+            if (responseTables.get(i).isGiverNameYou()) {
+                selfResponseTables.add(responseTables.get(i));
+            }
+        }
+        return selfResponseTables;
+    }
+
+    public List<FeedbackResultsResponseTable> getOthersResponseTables() {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java]: Same here.

File: src/main/webapp/WEB-INF/tags/student/feedbackResults/questionWithResponses.tag [hunk 5]
```diff
@@ -13,10 +13,20 @@
 
     <c:if test="${questionWithResponses.questionDetails.individualResponsesShownToStudents}">
 
-      <c:forEach items="${questionWithResponses.responseTables}" var="responseTable">
-        <feedbackResults:responseTable responseTable="${responseTable}"/>
+      <c:forEach items="${questionWithResponses.othersResponseTables}" var="othersResponseTable">
+        <feedbackResults:responseTable responseTable="${othersResponseTable}"/>
       </c:forEach>
 
+      <c:if test="${!questionWithResponses.selfResponseTablesEmpty}">
```
Comment (xpdavid) [inline src/main/webapp/WEB-INF/tags/student/feedbackResults/questionWithResponses.tag]: Personally, I prefer `isSelfResponseTablesEmpty`. It's more like a boolean variable. You can achieve this by setting getter `getIsSelfResponseTablesEmpty`

File: src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java [hunk 6]
```diff
@@ -1,5 +1,6 @@
 package teammates.ui.template;
 
+import java.util.ArrayList;
 import java.util.List;
 
 public class StudentFeedbackResultsQuestionWithResponses {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java]: If u don't mind. Can u also add a javadoc for this class. Like mentioning we separate the responses to self responses and others.

> We are begin to fix the missing javadoc now. It is also a good practice for u to dig into the code :)

**Comment by dalessr** [comment]
@joanneong @xpdavid Ready for review.

Some tips for the solution:
- Since `getResponseTables` method is separated by `getSelfResponsesTables` and `getOthersResponsesTables`, thus I added more tests in `StudentFeedbackResultsPageDataTest` to test on the performance of both methods.
- There is no javadoc for other classes in `ui/template` folder which I can refer to. Please take a look and correct me if the javadoc is written improperly.

**Comment by wkurniawan07** [comment]
@dalessr @xpdavid actually, any class in `ui.template` will not be very meaningful to be documented. Even when I said we need to strive towards fully documented code base, that package is one of those that I'd allow omission of documentation, until we're actually ready for it.

The reason is the classes in that package were created in a rather ad-hoc manner. In a very ideal situation, a class in `ui.template` should have a one-to-one mapping with something in `WEB-INF/tags` (if you're familiar with Angular 2+, think of it as the `.ts` and the `.html` of the same component), but clearly we're far from that right now. Of course, if that one-to-one mapping is satisfied, then documenting will be meaningful. Something like:
```java
package teammates.ui.template.SomeClass;

import ...;

/**
 * The view model for someComponent.tag.
 */
public class SomeClass {

    ...

}
```
In the same manner, a class in `ui.pagedata` should be one-to-one-mapped with some `.jsp` file. This one is much more closely followed.

File: src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java [hunk 7]
```diff
@@ -1,7 +1,12 @@
 package teammates.ui.template;
 
 import java.util.List;
+import java.util.stream.Collectors;
 
+/**
+ * Manages student feedback results questions details, self-responses tables and others-responses tables separately
```
Comment (xpdavid) [inline src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java]: Thanks for the input @wkurniawan07 .

As mentioned, we can put `View Model for student/feedbackResults/questionWithResponses.tag`. Your explanation is also fair enough and we can keep it.

File: src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java [hunk 8]
```diff
@@ -96,19 +96,35 @@ public void testAll() throws EntityDoesNotExistException {
         assertEquals("", questionBundle1.getQuestionDetails().getAdditionalInfo());
         assertEquals("", questionBundle2.getQuestionDetails().getAdditionalInfo());
 
-        assertNotNull(questionBundle1.getResponseTables());
-        assertNotNull(questionBundle2.getResponseTables());
+        assertNotNull(questionBundle1.getSelfResponseTables());
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java]: Add `______TS("test xxxx");` to state what we are testing here.

**Comment by dalessr** [comment]
Status: Ready for review

File: src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java [hunk 9]
```diff
@@ -1,7 +1,13 @@
...
 
 import java.util.List;
+import java.util.stream.Collectors;
 
+/**
+ * View Model for student/feedbackResults/questionWithResponses.tag
+ * Manages student feedback results questions details, self-responses tables and others-responses tables separately
+ * for a specific question.
+ */
 public class StudentFeedbackResultsQuestionWithResponses {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java]:
```
/**
* View Model for student/feedbackResults/questionWithResponses.tag.
*
* Manages student feedback results questions details, self-responses tables and others-responses tables separately
* for a specific question.
*/
```
The first line is the summary and should ends with `.`.
There should be a empty space between the summary and the detail description.

File: src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java [hunk 10]
```diff
@@ -96,19 +96,37 @@ public void testAll() throws EntityDoesNotExistException {
...
+        assertNotNull(questionBundle2.getSelfResponseTables());
+        assertNotNull(questionBundle2.getOthersResponseTables());
 
-        assertNotNull(questionBundle1.getResponseTables().get(0).getResponses());
+        assertTrue(questionBundle1.getSelfResponseTables().get(0).isGiverNameYou());
+        assertEquals("You", questionBundle1.getSelfResponseTables().get(0).getRecipientName());
 
-        assertEquals("You", questionBundle1.getResponseTables().get(0).getResponses()
+        assertEquals(0, questionBundle1.getOthersResponseTables().size());
+        assertNotNull(questionBundle1.getSelfResponseTables().get(0).getResponses());
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java]: - Any reason of putting these assertion together? One is testing getOthersResponseTables while the other is testing `getSelfResponseTables`.

File: src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java [hunk 11]
```diff
@@ -96,19 +96,37 @@ public void testAll() throws EntityDoesNotExistException {
...
 
-        assertEquals("Student 1 self feedback.", questionBundle1.getResponseTables().get(0).getResponses()
+        assertEquals("Student 1 self feedback.", questionBundle1.getSelfResponseTables().get(0).getResponses()
                                         .get(0).getAnswer());
 
+        assertFalse(questionBundle2.getOthersResponseTables().get(0).isGiverNameYou());
+        assertEquals("You", questionBundle2.getOthersResponseTables().get(0).getRecipientName());
+
+        assertEquals(1, questionBundle2.getSelfResponseTables().size());
+        assertNotNull(questionBundle2.getOthersResponseTables().get(0).getResponses());
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java]: Same here

File: src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java (hunk 9 above)
Comment (dalessr) [reply src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java]: As specified by the checkstyle, do I need to put a `<p>` before the description?

File: src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java (hunk 10 above)
Comment (dalessr) [reply src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java]: This is just because I was still following the original test cases, where the cases fall into two tables. Will reorder it.

**Comment by dalessr** [comment]
@joanneong @xpdavid Ready for review

File: src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java [hunk 12]
```diff
@@ -1,7 +1,14 @@
...
 
 import java.util.List;
+import java.util.stream.Collectors;
 
+/**
+ * View Model for student/feedbackResults/questionWithResponses.tag
+ *
+ * <p>Manages student feedback results questions details, self-responses tables and others-responses tables separately
+ * for a specific question.
+ */
```
Comment (joanneong) [inline src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java]: As @wkurniawan07 mentioned in his comment some time ago, it may not be very meaningful to add JavaDoc for code in `ui.template` right now (though I personally think it's fine here :))

One question here though - why did you add `<p>` in the JavaDoc comment?

File: src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java [hunk 13]
```diff
@@ -96,19 +96,33 @@ public void testAll() throws EntityDoesNotExistException {
...
+        assertEquals(1, questionBundle2.getSelfResponseTables().size());
+
+        assertFalse(questionBundle2.getOthersResponseTables().get(0).isGiverNameYou());
+        assertEquals("You", questionBundle2.getOthersResponseTables().get(0).getRecipientName());
+        assertNotNull(questionBundle2.getOthersResponseTables().get(0).getResponses());
+        assertEquals("student2 In Course1", questionBundle2.getOthersResponseTables().get(0).getResponses()
+                .get(0).getGiverName());
+        assertEquals("Response from student 2 to student 1.", questionBundle2.getOthersResponseTables().get(0)
+                .getResponses().get(0).getAnswer());
+
```
Comment (joanneong) [inline src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java]: Are there particular structures that you're following when you wrote the tests here? They feel somewhat haphazardly put together (to me)...I can get the rough sense that you're checking if the tables are correct and if the receiver and giver are displayed correctly but I find the way that the tests are written here a tad bit messy.

File: src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java (hunk 12 above)
Comment (dalessr) [reply src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java]: I checked the coding standard for JavaDoc, and it informs that this is the correct way of adding a new paragraph if there's an empty line above. Otherwise it cannot pass the travis test. If it really does't look nice, I think it should be ok to just delete the entire JavaDoc.

File: src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java (hunk 13 above)
Comment (dalessr) [reply src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java]: There are 2 sets of `questionBundle`s here, with different example responses added inside. The tests I added here is to check the correctness of recipient names and giver names in `selfResponseTables` as well as `othersResponseTables` for each of the `questionBundle`. The tests are actually similar to the original one, it's just that there are only tests for one table of the first `questionBundle` in the original test cases, while I added tests for two tables of both `questionBundle`.

File: src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java (hunk 12 above)
Comment (joanneong) [reply src/main/java/teammates/ui/template/StudentFeedbackResultsQuestionWithResponses.java]: Leave it there for now then, unless @xpdavid feels otherwise :P

File: src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java [hunk 14]
```diff
@@ -96,19 +96,33 @@ public void testAll() throws EntityDoesNotExistException {
...
 
-        assertNotNull(questionBundle1.getResponseTables());
-        assertNotNull(questionBundle2.getResponseTables());
+        ______TS("giver & receiver in self/others response tables");
 
-        assertEquals("You", questionBundle1.getResponseTables().get(0).getRecipientName());
+        assertNotNull(questionBundle1.getSelfResponseTables());
+        assertNotNull(questionBundle1.getOthersResponseTables());
+        assertNotNull(questionBundle2.getSelfResponseTables());
+        assertNotNull(questionBundle2.getOthersResponseTables());
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java]: U forget to test `getIsSelfResponseTablesEmpty`?

File: src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java [hunk 15]
```diff
@@ -96,19 +96,33 @@ public void testAll() throws EntityDoesNotExistException {
...
 
-        assertNotNull(questionBundle1.getResponseTables().get(0).getResponses());
-
-        assertEquals("You", questionBundle1.getResponseTables().get(0).getResponses()
+        assertTrue(questionBundle1.getSelfResponseTables().get(0).isGiverNameYou());
```
Comment (xpdavid) [inline src/test/java/teammates/test/cases/pagedata/StudentFeedbackResultsPageDataTest.java]: You can always add some comment to state what u are testing (like the one u explained to @joanneong )

A example here `// test that self response table contains correct response`

**Comment by dalessr** [comment]
@xpdavid @joanneong Ready for review

**Comment by dalessr** [comment]
@xpdavid @joanneong A gentle reminder to review this, thanks!
